### PR TITLE
Upgrade LLM module to support latest OpenAI connectors

### PR DESCRIPTION
**Pull request by nadheesh**
## Purpose
Fixes https://github.com/wso2-enterprise/choreo/issues/26577

### CODE REVIEW

File: ballerina/llm.bal [hunk 1]
```diff
@@ -48,18 +48,50 @@ public type ChatModelConfig readonly & record {|
...
+    string name?;
+|};
+
+# System chat message record
+public type ChatSystemMessage record {|
+    # Role of the message
+    SYSTEM role;
+    # Content of the message
+    string content;
+    # An optional name for the participant. Provides the model information to differentiate between participants of the same role.
```
Comment (SasinduDilshara) [inline ballerina/llm.bal:67]: in, L65, there is no `.` at the end of the comment in this line it contains.
Shall we use a single convention.

Shall we add a fullstop to the description in L61, no fullstop for field levelcomments

https://ballerina.io/learn/style-guide/annotations-documentation-and-comments/

File: ballerina/llm.bal [hunk 2]
```diff
@@ -48,18 +48,50 @@ public type ChatModelConfig readonly & record {|
...
+    # The contents of the assistant message. Required unless `tool_calls` or `function_call` is specified.
+    string? content?;
+    # An optional name for the participant. Provides the model information to differentiate between participants of the same role.
     string name?;
-    # Function call record if the message is a function call
+    # The function calls generated by the model, such as function calls.
     FunctionCall function_call?;
 |};
 
+// # Function message record
```
Comment (SasinduDilshara) [inline ballerina/llm.bal:83]:
```suggestion
# Function message record
```

File: ballerina/llm.bal [hunk 3]
```diff
@@ -280,8 +312,11 @@ public isolated class AzureChatGptModel {
         if response is error {
             return error LlmConnectionError("Error while connecting to the model", response);
         }
-        azure_chat:ChatCompletionResponseMessage? message = response.choices[0].message;
-        string? content = message?.content;
+        string? content = ();
+        record {|azure_chat:ChatCompletionResponseMessage message?; azure_chat:ContentFilterChoiceResults content_filter_results?; int index?; string finish_reason?; anydata...;|}[]? choices = response.choices;
```
Comment (SasinduDilshara) [inline ballerina/llm.bal:316]: This line exceed 120 characters. Shall we break into several lines.

File: ballerina/llm.bal [hunk 4]
```diff
@@ -302,16 +337,22 @@ public isolated class AzureChatGptModel {
...
-        if content is string {
-            return content;
-        }
-        azure_chat:ChatCompletionRequestMessage_function_call? 'function = message?.function_call;
-        if 'function is azure_chat:ChatCompletionRequestMessage_function_call {
-            return {
-                ...'function
-            };
+
+        record {|azure_chat:ChatCompletionResponseMessage message?; azure_chat:ContentFilterChoiceResults content_filter_results?; int index?; string finish_reason?; anydata...;|}[]? choices = response.choices;
```
[the remark this answers is not in the document]
Comment (nadheesh) [reply ballerina/llm.bal:341]: addressed

File: ballerina/llm.bal [hunk 5]
```diff
@@ -48,18 +48,50 @@ public type ChatModelConfig readonly & record {|
...
+# User chat message record
+public type ChatUserMessage record {|
     # Role of the message
-    ROLE role;
+    USER role;
     # Content of the message
-    string? content = ();
-    # Name of the function when the message is a function call
+    string content;
+    # An optional name for the participant. Provides the model information to differentiate between participants of the same role.
```
[the remark this answers is not in the document]
Comment (nadheesh) [reply ballerina/llm.bal:57]: addressed

File: ballerina/llm.bal (hunk 1 above)
Comment (nadheesh) [reply ballerina/llm.bal:67]: will address, but if we have multiple sentences for a single attribute, shouldn't we use `.` for each sentence. This is what I was told as the convention.

File: ballerina/llm.bal (hunk 2 above)
Comment (nadheesh) [reply ballerina/llm.bal]: fixed

File: ballerina/llm.bal (hunk 3 above)
Comment (nadheesh) [reply ballerina/llm.bal]: fixed

File: ballerina/llm.bal (hunk 1 above)
Comment (SasinduDilshara) [reply ballerina/llm.bal]: yes, +1

**Review by SasinduDilshara — APPROVED** [review]

File: ballerina/llm.bal [hunk 6]
```diff
@@ -232,8 +268,8 @@ public isolated class ChatGptModel {
         if content is string {
             return content;
         }
-        chat:ChatCompletionRequestMessage_function_call? 'function = message?.function_call;
-        if 'function is chat:ChatCompletionRequestMessage_function_call {
+        chat:ChatCompletionRequestAssistantMessage_function_call? 'function = message?.function_call;
```
Comment (SasinduDilshara) [inline ballerina/llm.bal:271]: I sthere specific reason to use `_` in these types?

File: ballerina/llm.bal [hunk 7]
```diff
@@ -280,8 +316,17 @@ public isolated class AzureChatGptModel {
...
+        string? content = ();
+        record {|
+            azure_chat:ChatCompletionResponseMessage message?;
+            azure_chat:ContentFilterChoiceResults content_filter_results?;
+            int index?;
+            string finish_reason?;
+            anydata...;
+        |}[]? choices = response.choices;
+        if choices !is () {
+            content = choices[0].message?.content;
```
Comment (SasinduDilshara) [inline ballerina/llm.bal:328]: There can be two error scenarios in here.
1. choices = ()
2. message = ().

Is it ok to have same error message for both?

Comment (nadheesh) [reply ballerina/llm.bal:328]: yes, because in either case we don't have expected answer from the LLM.

**Pull request merged**